typing/typecore.ml

```diff
@@ -7500,20 +7502,34 @@ and type_function
             (* Issue#12668: Retain type-directed disambiguation of
                ?x:(y : Variant.t = Constr)
             *)
-            let default =
-              match pat.ppat_desc with
-              | Ppat_constraint (_, Some sty, _) ->
-                  let gloc = { default.pexp_loc with loc_ghost = true } in
-                  Ast_helper.Exp.constraint_ default (Some sty) ~loc:gloc []
-              | _ -> default
+            let default, pat =
+              match arg_label' with
+              | Optional _ ->
+                  (match pat.ppat_desc with
+                  | Ppat_constraint (_, Some sty, _) ->
+                      let gloc = { default.pexp_loc with loc_ghost = true } in
+                      Ast_helper.Exp.constraint_ default (Some sty) ~loc:gloc []
+                      , pat
+                  | _ -> default, pat)
+              | Generic_optional _ ->
+                  (match extract_optional_tp_from_pattern_constraint pat with
+                  | None ->
+                      failwith
```

Comment:
`failwith` is ok here for now, but I think we should raise a proper error if the type is not `something option_like`.

Reply:
If this error is impossible at this stage, we should change the return type of `extract_optional_tp_from_pattern_constraint`.